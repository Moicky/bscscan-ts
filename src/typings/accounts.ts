import { APIResponse } from "./";

export type GetBalanceRequest = {
  address: string;
  tag?: "latest" | string;
};

export type GetBalanceResponse = APIResponse<string>;

export type GetBalanceMultiRequest = {
  address: string[];
  tag?: "latest" | string;
};

export type GetBalanceMultiResponse = APIResponse<Array<{ account: string; balance: string }>>;

export type GetTxListRequest = {
  address: string;
  startblock?: number;
  endblock?: number;
  sort?: "asc" | "desc";
  page?: number;
  offset?: number;
};

export type GetTxListResponse = APIResponse<{
  blockHash: string;
  blockNumber: string;
  confirmations: string;
  contractAddress: string;
  cumulativeGasUsed: string;
  from: string;
  gas: string;
  gasPrice: string;
  gasUsed: string;
  hash: string;
  input: string;
  isError: string;
  nonce: string;
  timeStamp: string;
  to: string;
  transactionIndex: string;
  txreceipt_status: string;
  value: string;
}>;

export type GetTxListInternalRequest = GetTxListRequest;

export type GetTxListInternalResponse = APIResponse<{
  blockNumber: string;
  contractAddress: string;
  errCode: string;
  from: string;
  gas: string;
  gasUsed: string;
  hash: string;
  input: string;
  isError: string;
  timeStamp: string;
  to: string;
  traceId: string;
  type: string;
  value: string;
}>;

export type GetTokenTxRequest = GetTxListRequest & {
  contractaddress?: string;
};

export type GetTokenTxResponse = APIResponse<
  Array<{
    blockHash: string;
    blockNumber: string;
    confirmations: string;
    contractAddress: string;
    cumulativeGasUsed: string;
    from: string;
    gas: string;
    gasPrice: string;
    gasUsed: string;
    hash: string;
    input: string;
    nonce: string;
    timeStamp: string;
    to: string;
    tokenDecimal: string;
    tokenName: string;
    tokenSymbol: string;
    transactionIndex: string;
    value: string;
  }>
>;

export type GetTokenNFTTxRequest = {
  address: string;
  blocktype?: "blocks" | string;
  page?: number;
  offset?: number;
};

export type GetTokenNFTTxResponse = APIResponse<{
  blockNumber: string;
  timeStamp: string;
  blockReward: string;
}>;
