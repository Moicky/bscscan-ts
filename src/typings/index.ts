export * from "./accounts";
export * from "./contracts";
export * from "./stats";
export * from "./transactions";

/**
 * Options used to initialize a BscScan client
 */
export type BscScanOptions = {
  apikey?: string;
  baseUrl?: string;
};

export type APIResponse<T> = {
  status: "1" | "0";
  message: string;
  result: T;
};

export type Token = {
  blockHash: string;
  blockNumber: string;
  confirmations: string;
  contractAddress: string;
  cumulativeGasUsed: string;
  from: string;
  gas: string;
  gasPrice: string;
  gasUsed: string;
  hash: string;
  input: string;
  nonce: string;
  timeStamp: string;
  to: string;
  tokenDecimal: string;
  tokenName: string;
  tokenSymbol: string;
  transactionIndex: string;
};

export type BEP20 = Token & { value: string };
export type ERC721 = Token & { tokenID: string };
