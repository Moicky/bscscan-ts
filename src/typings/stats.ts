import { APIResponse } from "./";

export type GetBNBTotalSupplyResponse = APIResponse<string>;

export type GetValidatorsResponse = APIResponse<
  Array<{
    validatorAddress: string;
    validatorName: string;
    validatorStatus: string;
    validatorVotingPower: string;
    validatorVotingPowerProportion: string;
  }>
>;

export type GetBNBLastPriceResponse = APIResponse<{
  ethbtc: string;
  ethbtc_timestamp: string;
  ethusd: string;
  ethusd_timestamp: string;
}>;
