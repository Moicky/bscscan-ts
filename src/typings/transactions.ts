export type GetTxReceiptStatusRequest = {
  // the string representing the transaction hash to check the execution status
  txhash: string;
};

export type GetTxReceiptStatusResponse = {
  // 0 for failed transactions and 1 for successful transactions.
  status: "0" | "1";
};
