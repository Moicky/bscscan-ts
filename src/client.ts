import { Accounts } from "./modules/accounts";
import { Contracts } from "./modules/contracts";
import { Stats } from "./modules/stats";
import { Transactions } from "./modules/transactions";
import { APIResponse, BscScanOptions } from "./typings";

const BASE_URL = "https://api.bscscan.com/api?";

class BscScan {
  apikey: string | undefined;
  baseUrl: string;

  accounts: Accounts;
  contracts: Contracts;
  transactions: Transactions;
  stats: Stats;

  constructor(opts: BscScanOptions) {
    const { apikey } = opts;

    this.apikey = apikey;
    this.baseUrl = opts.baseUrl || BASE_URL;

    this.accounts = new Accounts(this);
    this.contracts = new Contracts(this);
    this.stats = new Stats(this);
    this.transactions = new Transactions(this);
  }

  newRequest(method = "GET", module: string, action: string, opts: Record<string, any> = {}): Request {
    const params = new URLSearchParams({ module, action });

    if (this.apikey) {
      params.set("apikey", this.apikey);
    }

    for (const [key, value] of Object.entries(opts)) {
      if (Array.isArray(value)) {
        params.set(key, value.join(","));
      } else {
        params.set(key, value);
      }
    }

    const headers: Headers = new Headers();
    headers.set("Content-Type", "application/json; charset=UTF-8");

    const init: RequestInit = { method, headers };

    const url = new URL("/api?" + params.toString(), BASE_URL);

    const request = new Request(url.toString(), init);
    return request;
  }

  async query(module: string, action: string, opts: Record<string, any> = {}): Promise<any> {
    return this.do(this.newRequest("GET", module, action, opts));
  }

  async do<T>(request: Request): Promise<T> {
    const response: Response = await fetch(request);
    const responseBody: APIResponse<T> = await response.json();

    if (responseBody.status === "0") {
      throw new Error(responseBody.message + " - " + responseBody.result);
    }

    return responseBody.result;
  }
}

export { BscScan };
