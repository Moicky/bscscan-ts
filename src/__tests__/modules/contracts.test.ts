import "isomorphic-fetch";
import { BscScan } from "../../client";
import { VerifySourceCodeRequest } from "../../typings";
import { server } from "../server";
import { API_KEY } from "../setup";

describe("contracts", () => {
  const client = new BscScan({ apikey: API_KEY });

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should getAbi", async () => {
    const opts = { address: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82" };
    expect(await client.contracts.getAbi(opts)).toMatchSnapshot();
  });

  it("should getSourceCode", async () => {
    const opts = { address: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82" };
    expect(await client.contracts.getSourceCode(opts)).toMatchSnapshot();
  });

  it("should getCreatorAndTxnHash", async () => {
    const opts = {
      contractAddresses: [
        "0xa8c87a4b2de9bcebeadcc1ca4e6331fe28e3a378",
        "0xe9e7cea3dedca5984780bafc599bd69add087d56",
        "0xe4462eb568E2DFbb5b0cA2D3DbB1A35C9Aa98aad",
      ],
    };
    expect(await client.contracts.getCreatorAndTxnHash(opts)).toMatchSnapshot();
  });

  it("should verifySourceCode", async () => {
    const opts: VerifySourceCodeRequest = {
      contractAddress: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82",
      contractName: "CAKE",
      sourceCode: "",
      compilerVersion: "v0.6.12+commit.27d51765",
      optimizationUsed: "1",
      runs: 5000,
      licenseType: "None",
    };
    expect(await client.contracts.verifySourceCode(opts)).toMatchSnapshot();
  });

  it("should verifyProxyContract", async () => {
    const opts = {
      address: "0xbc46363a7669f6e12353fa95bb067aead3675c29",
      expectedImplementation: "0xe45a5176bc0f2c1198e2451c4e4501d4ed9b65a6",
    };
    expect(await client.contracts.verifyProxyContract(opts)).toMatchSnapshot();
  });

  it("should checkProxyVerification", async () => {
    const opts = { guid: "gwgrrnfy56zf6vc1fljuejwg6pelnc5yns6fg6y2i6zfpgzquz" };
    expect(await client.contracts.checkProxyVerification(opts)).toMatchSnapshot();
  });
});
