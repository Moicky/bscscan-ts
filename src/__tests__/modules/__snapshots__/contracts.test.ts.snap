// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contracts should checkProxyVerification 1`] = `"0xcbdcd3815b5f975e1a2c944a9b2cd1c985a1cb7f"`;

exports[`contracts should getAbi 1`] = `"[]"`;

exports[`contracts should getCreatorAndTxnHash 1`] = `
Array [
  Object {
    "contractAddress": "0x10e66c37b81858cf8377f766b6dfb5af6700a73c",
    "contractCreator": "0x187549f02d96d94945f2c4dd206cf58aeed8ebae",
    "txHash": "0x372eaf2d68524ebb0b5ec5b815983df2bcc29c9516a30116eb2bf02aae250c5e",
  },
  Object {
    "contractAddress": "0x10ed43c718714eb63d5aa57b78b54704e256024e",
    "contractCreator": "0xdb6f5fb9311ae8885620ee893887c3d85c8293d6",
    "txHash": "0x1bfbff8411ed44e609d911476b0d35a28284545b690902806ea0a7ff0453e931",
  },
  Object {
    "contractAddress": "0xa8c87a4b2de9bcebeadcc1ca4e6331fe28e3a378",
    "contractCreator": "0xd71110b333c547eb9139bef719d7ab4513107588",
    "txHash": "0xdd94fbce146d67c70ab6aea7fd0807706d1e1867ed67cc1b236516e6ba23c6f6",
  },
  Object {
    "contractAddress": "0xe9e7cea3dedca5984780bafc599bd69add087d56",
    "contractCreator": "0xf07c30e4cd6cfff525791b4b601bd345bded7f47",
    "txHash": "0x0f01fc521030f178115c880e200b09a40c9510f49de227aa880276f92670a3d6",
  },
]
`;

exports[`contracts should getSourceCode 1`] = `
Array [
  Object {
    "ABI": "[]",
    "CompilerVersion": "v0.6.12+commit.27d51765",
    "ConstructorArguments": "",
    "ContractName": "CakeToken",
    "EVMVersion": "Default",
    "Implementation": "",
    "Library": "",
    "LicenseType": "None",
    "OptimizationUsed": "1",
    "Proxy": "0",
    "Runs": "5000",
    "SourceCode": "pragma solidity 0.6.12;",
    "SwarmSource": "ipfs://9755240809e31aec9fa5790314682233ca431b7c4f252d7e4bba347e2e742086",
  },
]
`;

exports[`contracts should verifyProxyContract 1`] = `"a7xygt433w43um5fda3m6d9jk62jwtjhqdr9zcwj9fmfzcj14r"`;

exports[`contracts should verifySourceCode 1`] = `"a7xygt433w43um5fda3m6d9jk62jwtjhqdr9zcwj9fmfzcj14r"`;
